vss-processor/src/main/kotlin/org/eclipse/kuksa/vssprocessor/parser/json/JsonDefinitionParser.kt

```diff
@@ -0,0 +1,113 @@
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package org.eclipse.kuksa.vssprocessor.parser.json
+
+import com.google.gson.Gson
+import com.google.gson.JsonObject
+import com.google.gson.JsonParseException
+import org.eclipse.kuksa.vssprocessor.parser.VssDefinitionParser
+import org.eclipse.kuksa.vssprocessor.spec.VssSpecificationSpecModel
+import java.io.File
+import java.io.IOException
+
+private const val ROOT_KEY_VEHICLE = "Vehicle"
+
+private const val KEY_DATA_DESCRIPTION = "description"
+private const val KEY_DATA_TYPE = "type"
+private const val KEY_DATA_UUID = "uuid"
+private const val KEY_DATA_COMMENT = "comment"
+private const val KEY_DATA_DATATYPE = "datatype"
+private const val KEY_DATA_CHILDREN = "children"
+
+internal class JsonDefinitionParser : VssDefinitionParser {
+    private val dataKeys = listOf(
+        KEY_DATA_DESCRIPTION,
+        KEY_DATA_TYPE,
+        KEY_DATA_UUID,
+        KEY_DATA_COMMENT,
+        KEY_DATA_DATATYPE,
+        KEY_DATA_CHILDREN,
+    )
+
+    override fun parseSpecifications(definitionFile: File): List<VssSpecificationSpecModel> {
+        val vssSpecificationSpecModels = mutableListOf<VssSpecificationSpecModel>()
+
+        try {
+            val jsonStreamReader = definitionFile.reader()
+
+            val gson = Gson()
+            val rootJsonObject = gson.fromJson(jsonStreamReader, JsonObject::class.java)
+
+            if (rootJsonObject.has(ROOT_KEY_VEHICLE)) {
+                val vehicleJsonObject = rootJsonObject.getAsJsonObject(ROOT_KEY_VEHICLE)
+                vssSpecificationSpecModels += parseSpecModels(ROOT_KEY_VEHICLE, vehicleJsonObject)
+            } else {
+                throw IOException("Invalid VSS Specification file '${definitionFile.path}'")
+            }
+        } catch (e: JsonParseException) {
+            throw IOException("Invalid VSS Specification file '${definitionFile.path}'", e)
+        }
+
+        return vssSpecificationSpecModels.toList()
+    }
+
+    private fun parseSpecModels(
+        vssPath: String,
+        jsonObject: JsonObject,
+    ): MutableList<VssSpecificationSpecModel> {
+        val parsedSpecModels = mutableListOf<VssSpecificationSpecModel>()
+
+        val parsedSpecModel = parseSpecModel(jsonObject, vssPath)
+        parsedSpecModels += parsedSpecModel
+
+        if (jsonObject.has(KEY_DATA_CHILDREN)) {
+            val childrenJsonElement = jsonObject.getAsJsonObject(KEY_DATA_CHILDREN)
+
+            val filteredKeys = childrenJsonElement.asMap().keys
+                .filter { key -> !dataKeys.contains(key) }
+
+            filteredKeys.forEach { key ->
+                val childJsonElement = childrenJsonElement.getAsJsonObject(key)
+                val newVssPath = "$vssPath.$key"
+                // recursively go deeper in hierarchy and parse next element
+                parsedSpecModels += parseSpecModels(newVssPath, childJsonElement)
```

Comment:
I think the recursion saves us quite a lot of complexity, even though it gave a lot of headache during implementation and took quite some time to get it to work

Reply:
I would have done the same I think. :)